### PR TITLE
Images in DB

### PR DESCRIPTION
**Pull request by chadweimer**
Information about attached images is now stored in the DB. This allows for faster retrieval since it cuts out the search step on S3.

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/chadweimer/gomp/pull/30" target="_blank">ReviewNinja

File: models/image.go [hunk 1]
```diff
@@ -93,62 +167,144 @@ func (m *RecipeImageModel) Save(recipeID int64, name string, data []byte) error
...
+func (m *RecipeImageModel) DeleteAllTx(recipeID int64, tx *sqlx.Tx) error {
 	dirPath := getDirPathForRecipe(recipeID)
-	return m.upl.DeleteAll(dirPath)
+	err := m.upl.DeleteAll(dirPath)
+	if err != nil {
+		return err
+	}
+
+	_, err = tx.Exec(
+		"DELETE FROM recipe_note WHERE recipe_id = $1",
```
Comment (chadweimer) [inline models/image.go]: !fix This is using the wrong table name.

File: models/recipe.go [hunk 2]
```diff
@@ -193,50 +211,24 @@ func (m *RecipeModel) DeleteTx(id int64, tx *sqlx.Tx) error {
...
-	}
-
-	offset := count * (page - 1)
-	rows, err := m.db.Query(
-		"SELECT DISTINCT "+
-			"r.id, r.name, r.serving_size, r.nutrition_info, r.ingredients, r.directions, COALESCE(g.rating, 0) "+
-			"FROM recipe AS r LEFT OUTER JOIN recipe_rating AS g ON g.recipe_id = r.id "+
-			"ORDER BY r.name LIMIT $1 OFFSET $2",
-		count, offset)
+func (m *RecipeModel) listAllIds() ([]int64, error) {
```
Comment (chadweimer) [inline models/recipe.go]: This method really isn't needed. I could inline.

File: models/image.go [hunk 3]
```diff
@@ -93,62 +167,144 @@ func (m *RecipeImageModel) Save(recipeID int64, name string, data []byte) error
...
+	if err != nil {
+		return "", "", err
+	}
+
+	return origURL, thumbURL, nil
+}
+
+// ReadTx retrieves the information about the recipe from the database, if found,
+// using the specified transaction. If no recipe exists with the specified ID,
+// a NoRecordFound error is returned.
```
Comment (chadweimer) [inline models/image.go]: !fix This comment is referencing the wrong error name.

File: models/models.go [hunk 4]
```diff
@@ -49,28 +51,62 @@ func New(cfg *conf.Config) *Model {
...
+	newDbVersion, err := migrate.Version(databaseURL, migrationPath)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	return previousDbVersion, newDbVersion, nil
+}
+
+func (m *Model) postMigrate() error {
+	tx, err := m.db.Beginx()
```
Comment (chadweimer) [inline models/models.go]: This could be short-circuited when prev == current.

File: models/image.go (hunk 3 above)
Comment (chadweimer) [inline models/image.go]: !fixed

File: models/image.go (hunk 1 above)
Comment (chadweimer) [inline models/image.go]: !fixed

File: models/models.go (hunk 4 above)
Comment (chadweimer) [inline models/models.go]: !resolved

File: models/recipe.go (hunk 2 above)
Comment (chadweimer) [inline models/recipe.go]: !resolved

**Pull request merged**